File: arithmetic_coding.go
package main

import (
	"fmt"
	"strings"
)

func main() {
	var query string
	var number float64

	fmt.Println("=================")
	fmt.Print("文字列を入力: ")
	fmt.Scan(&query)
	num := str2num(query, 0.5, 0.3, 0.15, 0.05)
	fmt.Println(num)
	fmt.Println("=================")

	fmt.Print("算術符号を入力: ")
	fmt.Scan(&number)
	str := num2str(number, 0.5, 0.3, 0.15, 0.05)
	fmt.Println(str)
	fmt.Println("=================")
}

func str2num(query string, a float64, b float64, c float64, d float64) (float64) {
	front := map[string]float64{
		"a": 0.,
		"b": a,
		"c": a+b,
		"d": a+b+c,
	}

	var back = map[string]float64{
		"a": a,
		"b": a+b,
		"c": a+b+c,
		"d": a+b+c+d,
	}

	var lineWidth = map[string]float64{
		"a": a,
		"b": b,
		"c": c,
		"d": d,
	}
	queryArray := strings.Split(query, "")
	var value float64 = front[queryArray[0]]
	var width float64 = back[queryArray[0]]

	for i:=1; i<len(queryArray); i++ {
		value = value + width * front[queryArray[i]]
		width = width * lineWidth[queryArray[i]]
	}

	return value
}

func num2str(number float64, a float64, b float64, c float64, d float64) (string) {
	var front = map[string]float64{
		"a": 0.,
		"b": a,
		"c": a+b,
		"d": a+b+c,
	}

	var back = map[string]float64{
		"a": a,
		"b": a+b,
		"c": a+b+c,
		"d": a+b+c+d,
	}

	var lineWidth = map[string]float64{
		"a": a,
		"b": b,
		"c": c,
		"d": d,
	}
	
	var ans string = ""
	var letters []string = []string{"a", "b", "c", "d"}

	for number>0. {
		for i:=0; i<len(letters); i++ {
			if front[letters[i]] <= number && number < back[letters[i]] {
				ans += letters[i]
				number = (number - front[letters[i]]) / lineWidth[letters[i]]
			}
		}
	}
	
	return ans
}

File: arithmetic_coding.py
# -*- coding: utf-8 -*-

# 文字列から算術符号を計算
def str2int(query, a=0.5, b=0.3, c=0.15, d=0.05):
    # 各範囲の開始と終わりと幅を辞書で格納する
    front = {
        "a": 0,
        "b": a,
        "c": a+b,
        "d": a+b+c
    }
    back = {
        "a": a,
        "b": a+b,
        "c": a+b+c,
        "d": a+b+c+d
    }
    line_width = {
        "a": a,
        "b": back["b"]-a,
        "c": back["c"]-back["b"],
        "d": back["d"]-back["c"]
    }
    
    value, width = 0, 0
    for i, q in enumerate(query):
        if i == 0:
            value, width = front[q], back[q]
        else:
            value = value + (width * front[q])
            width = width * line_width[q]
    
    return round(value, 2)

# 算術符号から文字列をデコード
def int2str(number, a=0.5, b=0.3, c=0.15, d=0.05):
    # 各範囲の開始と終わりと幅を辞書で格納する
    front = {
        "a": 0,
        "b": a,
        "c": a+b,
        "d": a+b+c
    }
    back = {
        "a": a,
        "b": a+b,
        "c": a+b+c,
        "d": a+b+c+d
    }
    line_width = {
        "a": a,
        "b": back["b"]-a,
        "c": back["c"]-back["b"],
        "d": back["d"]-back["c"]
    }
        
    ans = ""
    while number > 0.:
        if front["a"] <= number < back["a"]:
            ans += "a"
            number = round((number - front["a"]) / line_width["a"], 2)
            
        if front["b"] <= number < back["b"]:
            ans += "b"
            number = round((number - front["b"]) / line_width["b"], 2)
        
        if front["c"] <= number < back["c"]:
            ans += "c"
            number = round((number - front["c"]) / line_width["c"], 2)
        
        if front["d"] <= number < back["d"]:
            ans += "d"
            number = round((number - front["d"]) / line_width["d"], 2)
            
    
    return ans

if __name__ == '__main__':
    print("===============")
    S = input("文字列を入力: ")
    # print("文字列 => 算術符号")
    print(f"出力: {str2int(S)}")
    
    print("===============")
    N = float(input("算術符号を入力: "))
    # print("算術符号 => 文字列")
    print(f"出力: {int2str(N)}")
    print("===============")